blog/_posts/2019-03-02-custom-primitiv-type.md

```diff
@@ -0,0 +1,383 @@
+
+function Base.transpose(x::BitBoard64)
+    t = and_int(xor_int(x, lshr_int(x, 12)), BitBoard64(0x0000f0f00000f0f0))
+    x = xor_int(x, xor_int(t, shl_int(t, 12)))
+    t = and_int(xor_int(x, lshr_int(x, 24)), BitBoard64(0x00000000ff00ff00))
+    x = xor_int(x, xor_int(t, shl_int(t, 24)))
+    x
+end
+
+function reflect_vertical(x::BitBoard64)
+    t = and_int(xor_int(x, lshr_int(x, 32)), BitBoard64(0x00000000ffffffff))
+    x = xor_int(x, xor_int(t, shl_int(t, 32)))
+    t = and_int(xor_int(x, lshr_int(x, 16)), BitBoard64(0x0000ffff0000ffff))
+    x = xor_int(x, xor_int(t, shl_int(t, 16)))
+    x
+end
+
+function reflect_horizontal(x::BitBoard64)
+    t = and_int(xor_int(x, lshr_int(x, 8)), BitBoard64(0x00ff00ff00ff00ff))
+    x = xor_int(x, xor_int(t, shl_int(t, 8)))
+    t = and_int(xor_int(x, lshr_int(x, 4)), BitBoard64(0x0f0f0f0f0f0f0f0f))
+    x = xor_int(x, xor_int(t, shl_int(t, 4)))
+    x
+end
+
+function Base.transpose(x::BitBoard128)
+    t = and_int(xor_int(x, lshr_int(x, 24)), BitBoard128(0x00000000ff00ff0000000000ff00ff00))
+    x = xor_int(x, xor_int(t, shl_int(t, 24)))
+    t = and_int(xor_int(x, lshr_int(x, 48)), BitBoard128(0x0000000000000000ffff0000ffff0000))
+    x = xor_int(x, xor_int(t, shl_int(t, 48)))
+    x
+end
+
+
+function reflect_vertical(x::BitBoard128)
+    t = and_int(xor_int(x, lshr_int(x, 64)), BitBoard128(0x0000000000000000ffffffffffffffff))
+    x = xor_int(x, xor_int(t, shl_int(t, 64)))
+    t = and_int(xor_int(x, lshr_int(x, 32)), BitBoard128(0x00000000ffffffff00000000ffffffff))
+    x = xor_int(x, xor_int(t, shl_int(t, 32)))
+    x
+end
+
+function reflect_horizontal(x::BitBoard128)
+    t = and_int(xor_int(x, lshr_int(x, 16)), BitBoard128(0x0000ffff0000ffff0000ffff0000ffff))
+    x = xor_int(x, xor_int(t, shl_int(t, 16)))
+    t = and_int(xor_int(x, lshr_int(x, 8)), BitBoard128(0x00ff00ff00ff00ff00ff00ff00ff00ff))
+    x = xor_int(x, xor_int(t, shl_int(t, 8)))
+    x
+end
+
+function rotate_right(b::AbstractBoard)
+    b = transpose(b)
+    b = reflect_horizontal(b)
+    b
+end
+
+function rotate_left(b::AbstractBoard)
+    b = transpose(b)
+    b = reflect_vertical(b)
+    b
+end
+
+GetRow(t::BitBoard128, r::Int)::UInt32 = and_int(trunc_int(UInt32, lshr_int(t, 32 * (3 - r))), 0xffffffff)
+GetRow(t::BitBoard64, r::Int)::UInt16 = and_int(trunc_int(UInt16, lshr_int(t, 16 * (3 - r))), 0xffff)
+
+function empty(b::AbstractBoard)::Vector{Int}
+    a = Vector{Int}()
+    for i in 0:15
+        if b(i) == 0
+            push!(a, i)
+        end
+    end
+    return a
+end
+```
+
+
+<a id="reference"></a>
+
+## Core.Intrinsics Reference
+
+reference for some basic function under the `Core.Intrinsics`.
+1.  Type construction
+    
+    1.  `Core.Intrinsics.bitcase(type, value)`
+    
+        `bitcase` told the LLVM to make a new same sized value of a primitive type instance `value`         and mark it as type `type`. The type of `value` must have the same bitsize as `type`.
```

Comment:
why these spaces between `value` and "and". Same applies to next lines

Reply:
Oh, I guess it's due to some problems in my editor, let me remove it.